README.md
# Moneris API

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Total Downloads][ico-downloads]][link-downloads]
 
## Requirements

PHP 5.6 and later

## Composer

To get started, install the package via the Composer package manager:

```php
composer require craigpaul/moneris-api
```

## Instantiation

Creating a new Moneris instance is quite easy and straightforward.

```php
use CraigPaul\Moneris\Moneris;

...

$id = 'store1';
$token = 'yesguy';

// optional
$params = [  
  'environment' => Moneris::ENV_TESTING, // default: Moneris::ENV_LIVE
  'avs' => true, // default: false
  'cvd' => true, // default: false
  'cof' => true, // default: false
];

$gateway = (new Moneris($id, $token, $params))->connect();
```

```php
use CraigPaul\Moneris\Moneris;

...

$id = 'store1';
$token = 'yesguy';

// optional
$params = [  
  'environment' => Moneris::ENV_TESTING, // default: Moneris::ENV_LIVE
  'avs' => true, // default: false
  'cvd' => true, // default: false
  'cof' => true, // default: false
];

$gateway = Moneris::create($id, $token, $params);
```

> **Note:** Please note that the Moneris store id and API token are always required to be passed to the Moneris constructor or static create method.

## Transactions

To make a purchase, preauth a card, void a transaction, etc. is very straightforward once you have your Gateway instantiated ([see above](#instantiation)).

### Purchase

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->purchase($params);
```

### Pre-Authorization

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->preauth($params);
```

### Capture (Pre-Authorization Completion)

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->preauth($params);

$response = $gateway->capture($response->transaction);
```

### Void (Purchase Correction)

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->purchase($params);

$response = $gateway->void($response->transaction);
```

### Refund

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->purchase($params);

$response = $gateway->refund($response->transaction);
```

### Card Verification

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->verify($params);
```

### CVD and AVS

To take advantage of the Card Verification Digits and/or Address Verification Service provided by Moneris, you need to tell Moneris that upon instantiation ([as shown above](#instantiation)).

When making a CVD secured purchase, pre-authorization or card verification, you need to pass the following parameter to the Gateway method you are utilizing.

```php
$params = [
    // `cvd` needs to be included in your transaction parameters.
    'cvd' => '111',
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => $this->visa,
    'expdate' => '2012',
];

$response = $gateway->verify($params); // could be purchase, preauth, etc.
```

When making an AVS secured purchase, pre-authorization or card verification, you need to pass the following parameters to the Gateway method you are utilizing.

```php
$params = [
    // `avs_*` keys need to be included in your transaction parameters.
    'avs_street_number' => '123',
    'avs_street_name' => 'Fake Street',
    'avs_zipcode' => 'X0X0X0',
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => $this->visa,
    'expdate' => '2012',
];

$response = $gateway->verify($params); // could be purchase, preauth, etc.
```

> Note: When making an AVS or CVD secured transaction, even if AVS or CVD fails, you will still have to void the transaction (DAMN MONERIS!). There are two easy ways around this.

Verify the card first. Using this method, there is one additional caveat (let me repeat it again...DAMN MONERIS!). Your verification transaction and purchase transaction must have different `order_id` parameters. One such solution could be to prepend an specific prefix to the front of verification order ids.

```php
$response = $gateway->verify($params);

if ($response->successful && !$response->failedAvs && !$response->failedCvd) {
    $response = $gateway->purchase($params);
    
    if ($response->successful) {
        $receipt = $response->receipt();
    } else {
        $errors = $response->errors;
    }
}
```

Void the transaction.

```php
$response = $gateway->purchase($params);

if ($response->successful && ($response->failedAvs || $response->failedCvd)) {
    $errors = $response->errors;
    $response = $gateway->void($response->transaction);
} elseif (!$response->successful) {
    $errors = $response->errors;
} else {
    $receipt = $response->receipt();
}
```

### Credential On File

The credential on file is part of the new Visa requirements to pass the CVD/CVV2 data for transactions. 

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'data_key' => $key,
    'payment_indicator' => 'U',
    'payment_information' => '2',
    'issuer_id' => $issuer_id // this is optional
];

$response = $vault->purchase($params); // could be purchase, preauth, etc.
```

## Vault

The Moneris Vault allows you create and maintain credit card profiles on the Moneris servers instead of your own. To access the Vault, you will need to have your instantiated Gateway ([see above](#instantiation)).

```php
$vault = $gateway->cards();
```

### Add a Card

> **Note:** The expiry passed into the credit card is in the format of YYMM as that is how Moneris accepts it. 

```php
use CraigPaul\Moneris\CreditCard;

...

$card = CreditCard::create('4242424242424242', '2012');

$response = $vault->add($card);
```

### Update a Card

In order to maintain your credit card profiles, Moneris will send back a unique key for the profile that will allow you to keep track in your own database. You can retrieve the key once you have received your Receipt ([see more below](#responses-and-receipts))

```php
$card = CreditCard::create('4242424242424242', '2012');

$response = $vault->add($card);
$key = $response->receipt()->read('key');

$card->expiry = '2112';

$response = $vault->update($key, $card);
```

### Delete a Card

```php
$card = CreditCard::create('4242424242424242', '2012');

$response = $vault->add($card);
$key = $response->receipt()->read('key');

$response = $vault->delete($key);
```

### Attaching a Customer

In order to sync your customer information with the credit cards stored in the Vault, we can attach a basic `Customer` object to the `CreditCard`.

#### Add a Card

```php
use CraigPaul\Moneris\Customer;

...

$params = [
    'id' => uniqid('customer-', true),
    'email' => 'example@email.com',
    'phone' => '555-555-5555',
    'note' => 'Customer note',
];

$customer = Customer::create($params);
$card = CreditCard::create('4242424242424242', '2012');
$card = $card->attach($customer);

$response = $vault->add($card);
```

#### Update a Card and Customer

```php
use CraigPaul\Moneris\Customer;

...

$params = [
    'id' => uniqid('customer-', true),
    'email' => 'example@email.com',
    'phone' => '555-555-5555',
    'note' => 'Customer note',
];

$customer = Customer::create($params);
$card = CreditCard::create('4242424242424242', '2012');
$card = $card->attach($customer);

$response = $vault->add($card);
$key = $response->receipt()->read('key');

$card->customer->email = 'example2@email.com';

$response = $vault->update($key, $card);
```

### Tokenize a Previous Transaction

Used to create a credit card profile based on a previous transaction.

```php
$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
];

$response = $gateway->purchase($params);

$response = $vault->tokenize($response->transaction);
```

### Peek Into The Vault

If you have the need to look up the masked credit card number you can peek into the Vault.

```php
$card = CreditCard::create('4242424242424242', '2012');

$response = $vault->add($card);
$key = $response->receipt()->read('key');

$response = $vault->peek($key);
$receipt = $response->receipt();

$masked = $receipt->read('data')['masked_pan'];
```

### Retrieve Expiring Cards

This transaction can be performed no more then 2 times on any given calendar day as per the Moneris API.

```php
$response = $vault->expiring();
```

### Transactions

Credit cards stored in the Moneris Vault have a slightly different flow for purchasing and pre-authorization. Any of the other transactions work exactly the same as shown above. 

#### Purchase

```php
$card = CreditCard::create('4242424242424242', '2012');

$response = $vault->add($card);
$key = $response->receipt()->read('key');

$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
    'data_key' => $key,
];

$response = $vault->purchase($params); // 
```

> **Note:** The Vault is used for the transaction here instead of the base Gateway object.

#### Pre-Authorization

```php
$card = CreditCard::create('4242424242424242', '2012');

$response = $vault->add($card);
$key = $response->receipt()->read('key');

$params = [
    'order_id' => uniqid('1234-56789', true),
    'amount' => '1.00',
    'credit_card' => '4242424242424242',
    'expiry_month' => '12',
    'expiry_year' => '20',
    'data_key' => $key,
];

$response = $vault->preauth($params); // 
```

## Responses and Receipts

The `Response` and `Receipt` objects allow you to understand how everything went with your API call. After a transaction returns from being processed the `Response` will get validated and return all the relevant information for you.

### Response

The information available to you on the `Response` object is as follows:

#### Errors

```php
$errors = $response->errors;
```

Any errors that might occur during your transaction will be returned in the following format for you. It is returned in this format to allow you to handle any translation logic in your own app by utilizing the unique `title` and `field` keys in each error.

```php
// The following example would be returned when you forget to set the `order_id` on your transaction. 

$errors = [
    [
        'field' => 'order_id',
        'code' => self::PARAMETER_NOT_SET, // 2
        'title' => 'not_set'
    ],
];
```

#### Status

```php
$status = $response->status;
```

The status will return a status code matching the appropriate error returned. See below for an example of the possible statuses returned.
 
```php
ERROR                    = -23;
INVALID_TRANSACTION_DATA = 0;

FAILED_ATTEMPT            = -1;
CREATE_TRANSACTION_RECORD = -2;
GLOBAL_ERROR_RECEIPT      = -3;

SYSTEM_UNAVAILABLE    = -14;
CARD_EXPIRED          = -15;
INVALID_CARD          = -16;
INSUFFICIENT_FUNDS    = -17;
PREAUTH_FULL          = -18;
DUPLICATE_TRANSACTION = -19;
DECLINED              = -20;
NOT_AUTHORIZED        = -21;
INVALID_EXPIRY_DATE   = -22;

CVD               = -4;
CVD_NO_MATCH      = -5;
CVD_NOT_PROCESSED = -6;
CVD_MISSING       = -7;
CVD_NOT_SUPPORTED = -8;

AVS             = -9;
AVS_POSTAL_CODE = -10;
AVS_ADDRESS     = -11;
AVS_NO_MATCH    = -12;
AVS_TIMEOUT     = -13;

POST_FRAUD = -22;
```

#### Success

```php
$success = $response->successful
```

The successful property simply lets you know if your transaction has been processed successfully.

### Receipt

The `Receipt` object is your record of any information relevant to your transaction you have submitted. To retrieve your receipt once you have a response see the following.

```php
$response = $gateway->purchase($params);

$receipt = $response->receipt();
```

Depending on that type of transaction, you will have different items on your `Receipt` that you have available to read.
 
```php
$amount = $receipt->read('amount');
```

For a full list of possible readable receipt items, see below.

```php
amount - The amount of the transaction. (string)
authorization - The authorization code for the transaction. (string)
avs_result - The avs result code for the transaction. (string)
card - The card type used for the transaction. (string)
code - The response code for the transaction. (string)
complete - Whether the transaction had completed correctly or not. (boolean)
cvd_result - The cvd result code.  (string)
data - The data related to the customer and card for the transaction. (array)
date - The date of the transaction. (string)
id - The Moneris id of the receipt. (string)
iso - The ISO code for the transaction. (string)
key - The data key used for vault transactions. (string)
message - Any relevant message provided for the transaction. (string)
reference - The reference number for the transaction. (string)
time - The time of the transaction. (string)
transaction - The Moneris id of the transaction. (string)
type - The transaction type. (string)
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [Craig Paul][link-author]
- [All Contributors][link-contributors]

## License

Moneris API is open-sourced software licensed under the [MIT license](https://opensource.org/licenses/MIT).

[ico-version]: https://img.shields.io/packagist/v/craigpaul/moneris-api.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/craigpaul/moneris-api/master.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/craigpaul/moneris-api.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/craigpaul/moneris-api
[link-travis]: https://travis-ci.org/craigpaul/moneris-api
[link-downloads]: https://packagist.org/packages/craigpaul/moneris-api
[link-author]: https://github.com/craigpaul
[link-contributors]: ../../contributors


CHANGELOG.md
#Moneris API Changelog

###0.8.1 (2020-05-05)

####Changed
- replace {} with \[\] for array reference [`a1addcc813`](https://github.com/craigpaul/moneris-api/commit/a1addcc813)

###0.8.0 (2019-09-19)

####Added
- add credential on file support [`4e099821ce`](https://github.com/craigpaul/moneris-api/commit/4e099821ce)

####Changed

- Changes phpunit versions [`023127ad9a`](https://github.com/craigpaul/moneris-api/commit/023127ad9a)
- Specifically calls the vendor bin file for phpunit [`87a18dff46`](https://github.com/craigpaul/moneris-api/commit/87a18dff46)
- Removes unnecessary package. Fixes styling on phpunit xml file [`5c74feb91d`](https://github.com/craigpaul/moneris-api/commit/5c74feb91d)
- Updates travis config to use a matrix so we can control the version of phpunit being downloaded [`62623ce6df`](https://github.com/craigpaul/moneris-api/commit/62623ce6df)
- Removes support for full card lookup as Moneris recommends against using this feature since it can cause PCI compliance headaches [`e2d0c14974`](https://github.com/craigpaul/moneris-api/commit/e2d0c14974)

###0.7.0 (2019-07-19)

####Changed

- Removes support for full card lookup as Moneris recommends against using this feature since it can cause PCI compliance headaches [`e2d0c14974`](https://github.com/craigpaul/moneris-api/commit/e2d0c14974)

###0.6.3 (2017-01-09)

####Changed

- Adds code style fixer. Updates code style. Updates md files [`1eda9ca3a6`](https://github.com/craigpaul/moneris-api/commit/1eda9ca3a6) 

###0.6.2 (2017-01-09)

####Changed

- Updates card_verification validation to not require amount [`714b1624ce`](https://github.com/craigpaul/moneris-api/commit/714b1624ce) 

###0.6.1 (2016-12-28)

####Changed

- Merge pull request #4 from tappleby/bugfix-payment-success [`cb3b75ed84`](https://github.com/craigpaul/moneris-api/commit/cb3b75ed84) 
- Updates transaction to reset errors array indices [`262617c3ac`](https://github.com/craigpaul/moneris-api/commit/262617c3ac) 
- Update successful payment logic, AVS + CVD can fail but the transaction will still be successful. [`bbd85b8063`](https://github.com/craigpaul/moneris-api/commit/bbd85b8063) 
- Merge remote-tracking branch 'upstream/master' [`0f4f6597a6`](https://github.com/craigpaul/moneris-api/commit/0f4f6597a6)

####Added

- Adds another caveat to docs [`14525afdce`](https://github.com/craigpaul/moneris-api/commit/14525afdce) 
- Adds caveat example to readme [`7c35112183`](https://github.com/craigpaul/moneris-api/commit/7c35112183) 

###0.6.0 (2016-11-27)

####Changed

- Updates readme [`735686c4d7`](https://github.com/craigpaul/moneris-api/commit/735686c4d7)
 
####Added

- Adds item attached to a purchase test [`b95b9fb382`](https://github.com/craigpaul/moneris-api/commit/b95b9fb382) 
- Adds ability to attach items to a purchase [`9fdb5ca317`](https://github.com/craigpaul/moneris-api/commit/9fdb5ca317)

###0.5.2 (2016-10-26)

####Changed

- Renames mocks to stubs [`78014491e2`](https://github.com/craigpaul/moneris-api/commit/78014491e2)
- Renames Stubs to mocks [`f6ce06e709`](https://github.com/craigpaul/moneris-api/commit/f6ce06e709)
- Updates expiring test to have a mocked response to overcome Moneris Api limits [`d3a813df34`](https://github.com/craigpaul/moneris-api/commit/d3a813df34)
- Removes old gitkeep files [`4d5a293a26`](https://github.com/craigpaul/moneris-api/commit/4d5a293a26)
- Autoloads stubs in development [`3db6228eda`](https://github.com/craigpaul/moneris-api/commit/3db6228eda)

####Added
   
- Adds mock handler helper function [`ac85c82adf`](https://github.com/craigpaul/moneris-api/commit/ac85c82adf) 
- Adds logic to get all ResolveData keys when more then one exists in the response [`c2393778da`](https://github.com/craigpaul/moneris-api/commit/c2393778da) 
- Adds docblock entry for transaction [`efed307cd7`](https://github.com/craigpaul/moneris-api/commit/efed307cd7)   
- Adds an expiring card stub [`d6c3cabb06`](https://github.com/craigpaul/moneris-api/commit/d6c3cabb06) 

###0.5.1 (2016-10-25)

####Changed
 
- Response:: => self:: [`61192d0b5e`](https://github.com/craigpaul/moneris-api/commit/61192d0b5e)  
- Fix test + sprintf formatting [`a88c97b7dc`](https://github.com/craigpaul/moneris-api/commit/a88c97b7dc)  
- Fix typo with expdate. [`aaf3f8dfd7`](https://github.com/craigpaul/moneris-api/commit/aaf3f8dfd7) 

####Added

- Add some extra handling for invalid cc / exp date where moneris doesnt return a status code. [`63bcbad7ef`](https://github.com/craigpaul/moneris-api/commit/63bcbad7ef)
- Add unit test for expdate conversion [`4c99e6f1ed`](https://github.com/craigpaul/moneris-api/commit/4c99e6f1ed)

###0.5.0 (2016-10-25)

####Changed

- Changes amount cast to string [`57b0a212e0`](https://github.com/craigpaul/moneris-api/commit/57b0a212e0)
- Updates customer tests to match new preparable setting [`6fa73e5961`](https://github.com/craigpaul/moneris-api/commit/6fa73e5961)
- Adjusts customer to use preparable instead of direct setting [`85ba6e9f66`](https://github.com/craigpaul/moneris-api/commit/85ba6e9f66)
- Extracts prepare method to preparable trait [`66deaddfd6`](https://github.com/craigpaul/moneris-api/commit/66deaddfd6)
- Extracts param appender and accounts for recursion [`57f71419e2`](https://github.com/craigpaul/moneris-api/commit/57f71419e2)

####Added

- Adds proper setter for customer data [`9da220ec66`](https://github.com/craigpaul/moneris-api/commit/9da220ec66) 
- Adds purchase with customer info tests [`fbff9e7f0d`](https://github.com/craigpaul/moneris-api/commit/fbff9e7f0d) 
- Adds vault customer info pre authorization test [`01846f8956`](https://github.com/craigpaul/moneris-api/commit/01846f8956) 
- Adds pre authorization test that includes customer info [`ca72ea6630`](https://github.com/craigpaul/moneris-api/commit/ca72ea6630)  
- Adds faker for tests [`ccea71a15d`](https://github.com/craigpaul/moneris-api/commit/ccea71a15d)   

src/Response.php
<?php

namespace CraigPaul\Moneris;

/**
 * CraigPaul\Moneris\Response
 *
 * @property array $errors
 * @property bool $failedAvs
 * @property bool $failedCvd
 * @property null|int $status
 * @property bool $successful
 * @property \CraigPaul\Moneris\Transaction $transaction
 */
class Response
{
    use Gettable, Settable;

    const ERROR                    = -23;
    const INVALID_TRANSACTION_DATA = 0;

    const FAILED_ATTEMPT            = -1;
    const CREATE_TRANSACTION_RECORD = -2;
    const GLOBAL_ERROR_RECEIPT      = -3;

    const SYSTEM_UNAVAILABLE    = -14;
    const CARD_EXPIRED          = -15;
    const INVALID_CARD          = -16;
    const INSUFFICIENT_FUNDS    = -17;
    const PREAUTH_FULL          = -18;
    const DUPLICATE_TRANSACTION = -19;
    const DECLINED              = -20;
    const NOT_AUTHORIZED        = -21;
    const INVALID_EXPIRY_DATE   = -22;

    const CVD               = -4;
    const CVD_NO_MATCH      = -5;
    const CVD_NOT_PROCESSED = -6;
    const CVD_MISSING       = -7;
    const CVD_NOT_SUPPORTED = -8;

    const AVS             = -9;
    const AVS_POSTAL_CODE = -10;
    const AVS_ADDRESS     = -11;
    const AVS_NO_MATCH    = -12;
    const AVS_TIMEOUT     = -13;

    const POST_FRAUD = -22;

    /**
     * Any errors that arise from processing a transaction.
     *
     * @var array
     */
    protected $errors = [];

    /**
     * Determine if we have failed Address Verification Service verification.
     *
     * @var bool
     */
    protected $failedAvs = false;

    /**
     * Determine if we have failed Card Validation Digits verification.
     *
     * @var bool
     */
    protected $failedCvd = false;

    /**
     * The status code.
     *
     * @var null|int
     */
    protected $status = null;

    /**
     * Determines whether the response was successful.
     *
     * @var bool
     */
    protected $successful = true;

    /**
     * @var \CraigPaul\Moneris\Transaction
     */
    protected $transaction;

    /**
     * Create a new Response instance.
     *
     * @param \CraigPaul\Moneris\Transaction $transaction
     */
    public function __construct(Transaction $transaction)
    {
        $this->transaction = $transaction;
    }

    /**
     * Create a new Response instance.
     *
     * @param \CraigPaul\Moneris\Transaction $transaction
     *
     * @return $this
     */
    public static function create(Transaction $transaction)
    {
        return new static($transaction);
    }

    /**
     * Retrieve the transaction's receipt if it is available.
     *
     * @return \CraigPaul\Moneris\Receipt|null
     */
    public function receipt()
    {
        if (!is_null($response = $this->transaction->response)) {
            return new Receipt($response->receipt);
        }

        return null;
    }

    /**
     * Validate the response.
     *
     * @return $this
     */
    public function validate()
    {
        $receipt = $this->receipt();
        $gateway = $this->transaction->gateway;

        if ($receipt->read('id') === 'Global Error Receipt') {
            $this->status = self::GLOBAL_ERROR_RECEIPT;
            $this->successful = false;

            return $this;
        }

        $this->successful = $receipt->successful();

        if (!$this->successful) {
            $this->status = $this->convertReceiptCodeToStatus($receipt);
            return $this;
        }

        $code = !is_null($receipt->read('avs_result')) ? $receipt->read('avs_result') : false;

        if ($gateway->avs && $code && $code !== 'null' && !in_array($code, $gateway->avsCodes)) {
            switch ($code) {
                case 'B':
                case 'C':
                    $this->status = self::AVS_POSTAL_CODE;
                    break;
                case 'G':
                case 'I':
                case 'P':
                case 'S':
                case 'U':
                case 'Z':
                    $this->status = self::AVS_ADDRESS;
                    break;
                case 'N':
                    $this->status = self::AVS_NO_MATCH;
                    break;
                case 'R':
                    $this->status = self::AVS_TIMEOUT;
                    break;
                default:
                    $this->status = self::AVS;
            }

            $this->failedAvs = true;

            return $this;
        }

        $code = !is_null($receipt->read('cvd_result')) ? $receipt->read('cvd_result') : null;

        if ($gateway->cvd && !is_null($code) && $code !== 'null' && !in_array($code[1], $gateway->cvdCodes)) {
            $this->status = self::CVD;
            $this->failedCvd = true;

            return $this;
        }

        return $this;
    }

    protected function convertReceiptCodeToStatus(Receipt $receipt)
    {
        $code = $receipt->read('code');

        if ($code === 'null' && $message_status = $this->convertReceiptMessageToStatus($receipt)) {
            $status = $message_status;
        } else {
            switch ($receipt->read('code')) {
                case '050':
                case '074':
                case 'null':
                    $status = self::SYSTEM_UNAVAILABLE;
                    break;
                case '051':
                case '482':
                case '484':
                    $status = self::CARD_EXPIRED;
                    break;
                case '075':
                    $status = self::INVALID_CARD;
                    break;

                case '208':
                case '475':
                    $status = self::INVALID_EXPIRY_DATE;
                    break;

                case '076':
                case '079':
                case '080':
                case '081':
                case '082':
                case '083':
                    $status = self::INSUFFICIENT_FUNDS;
                    break;
                case '077':
                    $status = self::PREAUTH_FULL;
                    break;
                case '078':
                    $status = self::DUPLICATE_TRANSACTION;
                    break;
                case '481':
                case '483':
                    $status = self::DECLINED;
                    break;
                case '485':
                    $status = self::NOT_AUTHORIZED;
                    break;
                case '486':
                case '487':
                case '489':
                case '490':
                    $status = self::CVD;
                    break;
                default:
                    $status = self::ERROR;
            }
        }

        return $status;
    }

    protected function convertReceiptMessageToStatus(Receipt $receipt)
    {
        $message = (string)$receipt->read('message');
        $status = null;

        if (preg_match('/invalid pan/i', $message)) {
            $status = self::INVALID_CARD;
        } elseif (preg_match('/invalid expiry date/i', $message)) {
            $status = self::INVALID_EXPIRY_DATE;
        }

        return $status;
    }
}
